spec/datadog/appsec/processor_spec.rb

```diff
@@ -0,0 +1,227 @@
+      let(:ruleset) { "#{__dir__}/../../../lib/datadog/appsec/assets/waf_rules/recommended.json" }
+
+      it { expect(described_class.new.send(:load_ruleset)).to be true }
+    end
+
+    context 'when ruleset is a non existing path' do
+      let(:ruleset) { '/does/not/exist' }
+
+      it { expect(described_class.new.send(:load_ruleset)).to be false }
+    end
+
+    context 'when ruleset is IO-like' do
+      let(:ruleset) { StringIO.new(JSON.dump(basic_ruleset)) }
+
+      it { expect(described_class.new.send(:load_ruleset)).to be true }
+    end
+
+    context 'when ruleset is Ruby' do
+      let(:ruleset) { basic_ruleset }
+
+      it { expect(described_class.new.send(:load_ruleset)).to be true }
+    end
+
+    context 'when ruleset is not parseable' do
+      let(:ruleset) { StringIO.new('this is not json') }
+
+      it { expect(described_class.new.send(:load_ruleset)).to be false }
+    end
+  end
+
+  describe '#create_waf_handle' do
+    let(:ruleset) { :recommended }
+
+    before do
+      allow(Datadog::AppSec.settings).to receive(:ruleset).and_return(ruleset)
+    end
+
+    context 'when ruleset is default' do
+      let(:ruleset) { :recommended }
+
+      it { expect(described_class.new.send(:create_waf_handle)).to be true }
+    end
+
+    context 'when ruleset is invalid' do
+      let(:ruleset) { { 'not' => 'valid' } }
+
+      it { expect(described_class.new.send(:create_waf_handle)).to be false }
+    end
+  end
+
+  describe '#initialize' do
+    let(:ruleset) { :recommended }
+
+    subject(:processor) { described_class.new }
+
+    before do
+      allow(Datadog::AppSec.settings).to receive(:ruleset).and_return(ruleset)
+    end
+
+    context 'when libddwaf fails to load' do
+      before do
+        expect(described_class).to receive(:require_libddwaf).and_return(false)
+
+        expect(Datadog.logger).to receive(:warn)
+      end
+
+      it { is_expected.to_not be_ready }
+    end
+
+    context 'when libddwaf fails to provide WAF' do
+      before do
+        expect(described_class).to receive(:require_libddwaf).and_return(true)
+        expect(described_class).to receive(:libddwaf_required?).and_return(false)
+
+        expect(Datadog.logger).to receive(:warn)
+      end
+
+      it { is_expected.to_not be_ready }
+    end
+
+    context 'when ruleset is a non existing path' do
+      let(:ruleset) { '/does/not/exist' }
+
+      before do
+        expect(Datadog.logger).to receive(:warn)
+      end
+
+      it { is_expected.to_not be_ready }
+    end
+
+    context 'when ruleset is not parseable' do
+      let(:ruleset) { StringIO.new('this is not json') }
+
+      before do
+        expect(Datadog.logger).to receive(:warn)
+      end
+
+      it { is_expected.to_not be_ready }
+    end
+
+    context 'when ruleset is invalid' do
+      let(:ruleset) { { 'not' => 'valid' } }
+
+      before do
+        expect(Datadog.logger).to receive(:warn)
+      end
+
+      it { is_expected.to_not be_ready }
+    end
```

Comment:
I would suggest simplifying this -- we already have tests that check above that `load_ruleset` returns a `true`/`false` for success, so here I think it makes sense to check that `#initialize` handles a `false` value, but since all `false` values are the same, testing them exhaustively doesn't seem to add a lot.

(Alternatively, we can also fully test `load_ruleset` entirely from `#initialize` and remove all of the direct tests)

Reply:
The idea here is to test various log message content  (that respect a certain RFC) being emitted on specific errors, an implementation I did not address yet to avoid scope creep. That will be for a separate PR.

Reply:
Fair enough. I do a bit worry that we're adding quite a bit of stuff speculatively because "it will be needed later". I'm more of a YAGNI person myself ;)

Reply:
I agree. To clarify, I Am Gonna Need It ;) as the log outputs are not just RFC-mandated but expected by system tests.